core/src/main/scala/akka/persistence/cassandra/journal/TagWriters.scala

```diff
@@ -30,18 +30,18 @@ import akka.util.ByteString
 @InternalApi private[akka] object TagWriters {
 
   private[akka] case class TagWritersSession(
-    tagWritePs:         Future[PreparedStatement],
-    tagWriteWithMetaPs: Future[PreparedStatement],
+    tagWritePs:         () => Future[PreparedStatement],
+    tagWriteWithMetaPs: () => Future[PreparedStatement],
     executeStatement:   Statement => Future[Done],
     selectStatement:    Statement => Future[ResultSet],
-    tagProgressPs:      Future[PreparedStatement],
-    tagScanningPs:      Future[PreparedStatement]) {
+    tagProgressPs:      () => Future[PreparedStatement],
+    tagScanningPs:      () => Future[PreparedStatement]) {
 
     def writeBatch(tag: Tag, events: Seq[(Serialized, Long)])(implicit ec: ExecutionContext): Future[Done] = {
       val batch = new BatchStatement(BatchStatement.Type.UNLOGGED)
       val tagWritePSs = for {
-        withMeta <- tagWriteWithMetaPs
-        withoutMeta <- tagWritePs
+        withMeta <- tagWriteWithMetaPs()
+        withoutMeta <- tagWritePs()
```

Comment:
Hmm, so we want it to potentially recreate the prepared statement for _every_ call to `writeBatch` or am I reading this wrong?

Reply:
it's cached in `CassandraSession`, which is aware of if the underlying sessions is initialized or failed

Once things are connected the Cassandra driver takes care of reconnects and such, but if the initial connect attempt failed the Future of the prepared statements would remain failed in the old code.

Reply:
Ah, ok, that caching in the session was the piece of the puzzle I was missing.